### PR TITLE
Add rotation matrix to node

### PR DESCRIPTION
**Pull request by ezrayst**
Node:
- [x] add function `assign_rotation_matrix` in `node` to store rotation matrix for transformation to global to local coordinate (and vice versa)
- Test for `rotation_matrix` in `node` will be done together when applying boundary conditions just like testing for velocity constraints.

Mesh:
- [x] in `mesh`, need to make `assign_rotation_matrix` which will include the call to `geometry`'s `rotation_matrix`
- [x] make test cases for `assign_rotation_matrix` in `mesh`

### CODE REVIEW

File: include/mesh.h [hunk 1]
```diff
@@ -206,6 +207,12 @@ class Mesh {
       const std::vector<std::tuple<mpm::Index, unsigned, double>>&
           velocity_constraints);
 
+  //! Assign rotation matrix
+  //! \param[in] inclined_velocity_constraints Constraint at node id and euler
```
Comment (kks32) [inline include/mesh.h]: Your input param doesn't match the function argument

File: include/mesh.h [hunk 2]
```diff
@@ -206,6 +207,12 @@ class Mesh {
...
+  //! angles
+  bool assign_rotation_matrices(
```
Comment (kks32) [inline include/mesh.h]: Shouldn't this be `compute_nodal_rotation_matrices`

File: include/mesh.tcc [hunk 3]
```diff
@@ -7,6 +7,8 @@ mpm::Mesh<Tdim>::Mesh(unsigned id, bool isoparametric)
   //! Logger
   std::string logger = "mesh::" + std::to_string(id);
   console_ = std::make_unique<spdlog::logger>(logger, mpm::stdout_sink);
+  //! Geometry
+  geometry_ = std::make_unique<mpm::Geometry<Tdim>>();
```
Comment (kks32) [inline include/mesh.tcc:11]: Why is geometry a class?

File: include/mesh.tcc [hunk 4]
```diff
@@ -519,6 +521,41 @@ bool mpm::Mesh<Tdim>::assign_particles_volumes(
...
+  try {
+    if (!nodes_.size())
+      throw std::runtime_error(
+          "No nodes have been assigned in mesh, cannot assign rotation "
+          "matrix");
+
+    // Loop through nodal_euler_angles of different nodes
+    for (const auto& nodal_euler_angles : euler_angles) {
+      // Node id
+      mpm::Index nid = std::get<0>(nodal_euler_angles);
```
Comment (kks32) [inline include/mesh.tcc]: Why is this still a tuple access?

File: include/node.h [hunk 5]
```diff
@@ -172,6 +172,11 @@ class Node : public NodeBase<Tdim> {
   //! Apply velocity constraints
   void apply_velocity_constraints() override;
 
+  //! Assign rotation matrix
+  //! \param[in] rotation_matrix Rotation matrix of the node
+  bool assign_rotation_matrix(
+      Eigen::Matrix<double, Tdim, Tdim> rotation_matrix) override;
```
Comment (kks32) [inline include/node.h]: This should be a `const` reference

File: include/node.h [hunk 6]
```diff
@@ -172,6 +172,11 @@ class Node : public NodeBase<Tdim> {
   //! Apply velocity constraints
   void apply_velocity_constraints() override;
 
+  //! Assign rotation matrix
+  //! \param[in] rotation_matrix Rotation matrix of the node
+  bool assign_rotation_matrix(
```
Comment (kks32) [inline include/node.h]: Why is it a `bool` if it is manually set to true for all cases?

File: include/node.h (hunk 5 above)
Comment (kks32) [inline include/node.h]: Can we just assign it here directly?

File: include/node.tcc [hunk 7]
```diff
@@ -276,3 +276,15 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_velocity_constraints() {
     this->acceleration_(direction, phase) = 0.;
   }
 }
+
```
Comment (kks32) [inline include/node.tcc]: Remove this section

File: tests/mesh_test.cc [hunk 8]
```diff
@@ -821,6 +822,31 @@ TEST_CASE("Mesh is checked for 2D case", "[mesh][2D]") {
                   false);
         }
 
+        // Test assign rotation matrices to nodes
+        SECTION("Check assign rotation matrices to nodes") {
+          // Map of nodal id and euler angles
+          std::map<mpm::Index, Eigen::Matrix<double, Dim, 1>> euler_angles;
+          // Euler angles
+          euler_angles.emplace(std::make_pair(
```
Comment (kks32) [inline tests/mesh_test.cc:833]: Here you are assigning only 1 Euler angle per node?

File: tests/mesh_test.cc [hunk 9]
```diff
@@ -1747,6 +1773,31 @@ TEST_CASE("Mesh is checked for 3D case", "[mesh][3D]") {
...
+              2, (Eigen::Matrix<double, Dim, 1>() << 70. * M_PI / 180,
+                  80. * M_PI / 180, 90. * M_PI / 180)
+                     .finished()));
+          euler_angles.emplace(std::make_pair(
+              3, (Eigen::Matrix<double, Dim, 1>() << 100. * M_PI / 180,
+                  110. * M_PI / 180, 120. * M_PI / 180)
+                     .finished()));
+
+          REQUIRE(mesh->assign_rotation_matrices(euler_angles) == true);
+        }
```
Comment (kks32) [inline tests/mesh_test.cc:1843]: Where do you check for failures?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/mesh.tcc (hunk 3 above)
Comment (ezrayst) [reply include/mesh.tcc:11]: We want to leave equations about geometry within a class by itself, right?

File: tests/mesh_test.cc (hunk 8 above)
Comment (ezrayst) [reply tests/mesh_test.cc:833]: no, 2 or 3 depending on dimension

File: include/node.h (hunk 6 above)
Comment (ezrayst) [reply include/node.h]: It will always assign, so `void` then?

File: tests/mesh_test.cc (hunk 9 above)
Comment (ezrayst) [reply tests/mesh_test.cc:1843]: When does it fail though?

File: include/mesh.tcc (hunk 3 above)
Comment (kks32) [reply include/mesh.tcc:11]: It doesn't have to be a class, there is no reason why `geometry` should be a class

File: include/node.h (hunk 6 above)
Comment (kks32) [reply include/node.h]: Yes

File: tests/mesh_test.cc (hunk 8 above)
Comment (kks32) [reply tests/mesh_test.cc:833]: Isn't this section Dim 2?

File: tests/mesh_test.cc (hunk 9 above)
Comment (kks32) [reply tests/mesh_test.cc:1843]: all the throws you do?

File: tests/mesh_test.cc (hunk 8 above)
Comment (ezrayst) [reply tests/mesh_test.cc:833]: Yes it is Dim 2, and we have two angles right?

File: include/mesh.tcc (hunk 3 above)
Comment (ezrayst) [reply include/mesh.tcc:11]: How do you want to change it then? Maybe make the needed functions in `geometry` to be in `mesh`?
The geometry class is not created in this PR and we agreed to have them this way last time.

File: tests/mesh_test.cc (hunk 9 above)
Comment (ezrayst) [reply tests/mesh_test.cc:1843]: For that we have to make another mesh with no nodes. I will make them.

File: include/mesh.h [hunk 10]
```diff
@@ -206,6 +207,11 @@ class Mesh {
       const std::vector<std::tuple<mpm::Index, unsigned, double>>&
           velocity_constraints);
 
+  //! Assign rotation matrix
```
Comment (kks32) [inline include/mesh.h]: compute and assign rotation matrix to nodes

File: include/mesh.h [hunk 11]
```diff
@@ -206,6 +207,11 @@ class Mesh {
...
+  //! \param[in] euler_angles Map of node numebr and respective euler_angles
```
Comment (kks32) [inline include/mesh.h]:
```suggestion
  //! \param[in] euler_angles Map of node number and respective euler_angles
```

File: include/mesh.tcc [hunk 12]
```diff
@@ -519,6 +521,38 @@ bool mpm::Mesh<Tdim>::assign_particles_volumes(
   return status;
 }
 
+//! Assign rotation matrix to nodes
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
//! Compute and assign rotation matrix to nodes
```

File: include/mesh.tcc [hunk 13]
```diff
@@ -519,6 +521,38 @@ bool mpm::Mesh<Tdim>::assign_particles_volumes(
...
+      // Node id
+      mpm::Index nid = nodal_euler_angles.first;
+      // Euler angles
+      Eigen::Matrix<double, Tdim, 1> angles = nodal_euler_angles.second;
+      // Compute rotation matrix
+      Eigen::Matrix<double, Tdim, Tdim> rotation_matrix =
+          geometry_->rotation_matrix(angles);
+
+      // Apply rotation matrix to nodes
+      map_nodes_[nid]->assign_rotation_matrix(rotation_matrix);
```
Comment (kks32) [inline include/mesh.tcc:545]:
```suggestion
      map_nodes_.at(nid)->assign_rotation_matrix(rotation_matrix);
```

File: include/node.h [hunk 14]
```diff
@@ -172,6 +172,13 @@ class Node : public NodeBase<Tdim> {
   //! Apply velocity constraints
   void apply_velocity_constraints() override;
 
+  //! Assign rotation matrix
+  //! \param[in] rotation_matrix Rotation matrix of the node
+  void assign_rotation_matrix(
+      const Eigen::Matrix<double, Tdim, Tdim> rotation_matrix) override {
```
Comment (kks32) [inline include/node.h]:
```suggestion
      const Eigen::Matrix<double, Tdim, Tdim>& rotation_matrix) override {
```

File: include/node_base.h [hunk 15]
```diff
@@ -152,6 +152,11 @@ class NodeBase {
   //! Apply velocity constraints
   virtual void apply_velocity_constraints() = 0;
 
+  //! Assign rotation matrix
+  //! \param[in] rotation_matrix Rotation matrix of the node
+  virtual void assign_rotation_matrix(
+      Eigen::Matrix<double, Tdim, Tdim> rotation_matrix) = 0;
```
Comment (kks32) [inline include/node_base.h]:
```suggestion
      Eigen::Matrix<double, Tdim, Tdim>& rotation_matrix) = 0;
```

File: tests/mesh_test.cc [hunk 16]
```diff
@@ -821,6 +822,35 @@ TEST_CASE("Mesh is checked for 2D case", "[mesh][2D]") {
...
+              2, (Eigen::Matrix<double, Dim, 1>() << 50. * M_PI / 180,
+                  60. * M_PI / 180)
+                     .finished()));
+          euler_angles.emplace(std::make_pair(
+              3, (Eigen::Matrix<double, Dim, 1>() << 70. * M_PI / 180,
+                  80. * M_PI / 180)
+                     .finished()));
+
+          REQUIRE(mesh->compute_nodal_rotation_matrices(euler_angles) == true);
+          // Check for failure
```
Comment (kks32) [inline tests/mesh_test.cc]:
```suggestion
          // Check for failure when no nodes are assigned
```

File: tests/mesh_test.cc [hunk 17]
```diff
@@ -821,6 +822,35 @@ TEST_CASE("Mesh is checked for 2D case", "[mesh][2D]") {
...
+          auto mesh_fail = std::make_shared<mpm::Mesh<Dim>>(1);
+          REQUIRE(mesh_fail->compute_nodal_rotation_matrices(euler_angles) ==
```
Comment (kks32) [inline tests/mesh_test.cc:868]: Check for missing node ids.

Comment (kks32) [inline tests/mesh_test.cc:868]: Check for empty `euler_angles`

File: tests/mesh_test.cc [hunk 18]
```diff
@@ -821,6 +822,35 @@ TEST_CASE("Mesh is checked for 2D case", "[mesh][2D]") {
...
+                  false);
```
Comment (kks32) [inline tests/mesh_test.cc:869]: We need to check for valid values of Euler angles, but Shyamini and Krishna will check this.

File: include/node.h [hunk 19]
```diff
@@ -199,6 +206,8 @@ class Node : public NodeBase<Tdim> {
   Eigen::Matrix<double, Tdim, Tnphases> acceleration_;
   //! Velocity constraints
   std::map<unsigned, double> velocity_constraints_;
+  //! Rotation matrix for inclined velocity constraints
+  Eigen::Matrix<double, Tdim, Tdim> rotation_matrix_;
   //! Logger
```
Comment (kks32) [inline include/node.h:215]:
```
//! A general velocity (non-Cartesian/inclined) constraint is specified at the node
bool generic_velocity_constraints_{false};
```

File: include/node.h [hunk 20]
```diff
@@ -172,6 +172,13 @@ class Node : public NodeBase<Tdim> {
   //! Apply velocity constraints
   void apply_velocity_constraints() override;
 
+  //! Assign rotation matrix
+  //! \param[in] rotation_matrix Rotation matrix of the node
+  void assign_rotation_matrix(
+      const Eigen::Matrix<double, Tdim, Tdim> rotation_matrix) override {
+    rotation_matrix_ = rotation_matrix;
```
Comment (kks32) [inline include/node.h:179]: `generic_velocity_constraints_ = true;`

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/mesh.tcc [hunk 21]
```diff
@@ -519,6 +521,38 @@ bool mpm::Mesh<Tdim>::assign_particles_volumes(
...
+template <unsigned Tdim>
+bool mpm::Mesh<Tdim>::compute_nodal_rotation_matrices(
+    const std::map<mpm::Index, Eigen::Matrix<double, Tdim, 1>>& euler_angles) {
+  bool status = false;
+  try {
+    if (!nodes_.size())
+      throw std::runtime_error(
+          "No nodes have been assigned in mesh, cannot assign rotation "
+          "matrix");
+
```
Comment (kks32) [inline include/mesh.tcc:534]: Check for the euler_angles size

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/mesh.tcc [hunk 22]
```diff
@@ -519,6 +521,38 @@ bool mpm::Mesh<Tdim>::assign_particles_volumes(
...
+    // Loop through nodal_euler_angles of different nodes
+    for (const auto& nodal_euler_angles : euler_angles) {
+      // Node id
+      mpm::Index nid = nodal_euler_angles.first;
+      // Euler angles
+      Eigen::Matrix<double, Tdim, 1> angles = nodal_euler_angles.second;
+      // Compute rotation matrix
+      Eigen::Matrix<double, Tdim, Tdim> rotation_matrix =
```
Comment (kks32) [inline include/mesh.tcc]: auto

File: tests/mesh_test.cc [hunk 23]
```diff
@@ -821,6 +822,35 @@ TEST_CASE("Mesh is checked for 2D case", "[mesh][2D]") {
                   false);
         }
 
+        // Test assign rotation matrices to nodes
+        SECTION("Check assign rotation matrices to nodes") {
+          // Map of nodal id and euler angles
+          std::map<mpm::Index, Eigen::Matrix<double, Dim, 1>> euler_angles;
+          // Euler angles
+          euler_angles.emplace(std::make_pair(
```
Comment (kks32) [inline tests/mesh_test.cc:833]: //! Node 0, Euler angles 10 and 20*.

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/mesh.tcc (hunk 13 above)
Comment (ezrayst) [reply include/mesh.tcc:545]: I think `map` defines `operator[]` to `.at()`. I think we should use `operator[]` in this case.

File: tests/mesh_test.cc [hunk 24]
```diff
@@ -826,7 +826,10 @@ TEST_CASE("Mesh is checked for 2D case", "[mesh][2D]") {
         SECTION("Check assign rotation matrices to nodes") {
           // Map of nodal id and euler angles
           std::map<mpm::Index, Eigen::Matrix<double, Dim, 1>> euler_angles;
-          // Euler angles
+          // Node 0 (non-existent) with Euler angles of 10 and 20  deg
```
Comment (kks32) [inline tests/mesh_test.cc:829]: What do you mean by non-existent? if so why does it succeed?

File: tests/mesh_test.cc [hunk 25]
```diff
@@ -826,7 +826,10 @@ TEST_CASE("Mesh is checked for 2D case", "[mesh][2D]") {
...
+          // Node 1 (non-existent) with Euler angles of 30 and 40 deg
```
Comment (kks32) [inline tests/mesh_test.cc:830]: Could you add comments to relevant sections?

File: tests/mesh_test.cc [hunk 26]
```diff
@@ -1781,7 +1799,10 @@ TEST_CASE("Mesh is checked for 3D case", "[mesh][3D]") {
         SECTION("Check assign rotation matrices to nodes") {
           // Map of nodal id and euler angles
           std::map<mpm::Index, Eigen::Matrix<double, Dim, 1>> euler_angles;
-          // Euler angles
+          // Node 0 (non-existent) with Euler angles of 10, 20 and 30 deg
```
Comment (kks32) [inline tests/mesh_test.cc:1802]: Why are these `(non-existent)`?

File: include/mesh.tcc [hunk 27]
```diff
@@ -539,11 +539,10 @@ bool mpm::Mesh<Tdim>::compute_nodal_rotation_matrices(
       // Euler angles
       Eigen::Matrix<double, Tdim, 1> angles = nodal_euler_angles.second;
       // Compute rotation matrix
-      Eigen::Matrix<double, Tdim, Tdim> rotation_matrix =
-          geometry_->rotation_matrix(angles);
+      auto rotation_matrix = geometry_->rotation_matrix(angles);
```
Comment (kks32) [inline include/mesh.tcc:542]: `const auto`

**Review by kks32 — CHANGES_REQUESTED** [review]

File: tests/mesh_test.cc (hunk 24 above)
Comment (ezrayst) [reply tests/mesh_test.cc:829]: Typo, sorry
